### PR TITLE
Add support for aggregations over distinct inputs

### PR DESCRIPTION
**Pull request by mbasmanova**
Add support for queries like SELECT count(distinct x) FROM t.

Inputs are first accumulated into SetAccumulators. When all input has been
received, unique values are aggregated.

### CODE REVIEW

File: velox/exec/GroupingSet.cpp [hunk 1]
```diff
@@ -227,12 +248,24 @@ void GroupingSet::addInputForActiveRows(
       continue;
     }
 
+    const auto& rows = getSelectivityVector(i);
+
+    if (aggregates_[i].distinct) {
+      if (!newGroups.empty()) {
```
Comment (xiaoxmeng) [inline velox/exec/GroupingSet.cpp:254]:
```
VELOX_CHECK_NOT_NULL(aggregates_[i])
```

File: velox/exec/GroupingSet.cpp [hunk 2]
```diff
@@ -300,6 +333,12 @@ std::vector<Accumulator> GroupingSet::accumulators() {
   if (sortedAggregations_ != nullptr) {
     accumulators.push_back(sortedAggregations_->accumulator());
   }
```
Comment (xiaoxmeng) [inline velox/exec/GroupingSet.cpp:335]: NYC: at l330: shall we add explicit to  Accumulator(Aggregate* aggregate) to enforce an explicit conversion from Aggregator* to Accumulator to ease code reading? Thanks!

File: velox/exec/DistinctAggregations.h [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+#include "velox/exec/Aggregate.h"
+#include "velox/exec/AggregateInfo.h"
+#include "velox/exec/RowContainer.h"
+#include "velox/vector/ComplexVector.h"
+
+namespace facebook::velox::exec {
+
+/// Computes aggregations over de-duplicated inputs. Supports aggregations with
+/// single input only.
```
Comment (xiaoxmeng) [inline velox/exec/DistinctAggregations.h:26]:
```
nit: single input column only?
```

File: velox/exec/DistinctAggregations.cpp [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+  column_index_t input;
+
+  VELOX_CHECK(!aggregates.empty());
+  for (auto i = 0; i < aggregates.size(); ++i) {
+    auto* aggregate = aggregates[i];
+    VELOX_CHECK_EQ(aggregate->inputs.size(), 1);
+    if (i == 0) {
+      input = aggregate->inputs[0];
+    } else {
+      VELOX_CHECK_EQ(input, aggregate->inputs[0]);
```
Comment (xiaoxmeng) [inline velox/exec/DistinctAggregations.cpp:154]: Not sure if we shall use USER_CHECK here or not? Thanks!

File: velox/exec/DistinctAggregations.cpp [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+template <typename T>
+class TypedDistinctAggregations : public DistinctAggregations {
+ public:
+  TypedDistinctAggregations(
+      std::vector<AggregateInfo*> aggregates,
+      const RowTypePtr& inputType,
+      memory::MemoryPool* pool)
+      : aggregates_{std::move(aggregates)}, pool_{pool} {
+    input_ = aggregates_[0]->inputs[0];
```
Comment (xiaoxmeng) [inline velox/exec/DistinctAggregations.cpp:31]: nit: input_ and inputType_  can be const members?

File: velox/exec/DistinctAggregations.cpp [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+      // Release memory back to HashStringAllocator to allow next
+      // aggregate to re-use it.
+      aggregate.function->destroy(groups);
+    }
+  }
+
+ private:
+  const std::vector<AggregateInfo*> aggregates_;
+  memory::MemoryPool* pool_;
```
Comment (xiaoxmeng) [inline velox/exec/DistinctAggregations.cpp:130]:
```
nit:  memory::MemoryPool* const pool_;
```

File: velox/exec/DistinctAggregations.cpp [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+    input_ = aggregates_[0]->inputs[0];
+    inputType_ = inputType->childAt(input_);
+  }
+
+  using AccumulatorType = aggregate::prestosql::SetAccumulator<T>;
+
+  /// Returns metadata about the accumulator used to store unique inputs.
+  Accumulator accumulator() const override {
+    return {
+        false, // isFixedSize
```
Comment (xiaoxmeng) [inline velox/exec/DistinctAggregations.cpp:40]: Nice comments!

**Review by xiaoxmeng — APPROVED** [review]
@mbasmanova Looks great. Thanks!

File: velox/exec/GroupingSet.cpp (hunk 2 above)
Comment (mbasmanova) [reply velox/exec/GroupingSet.cpp:335]: I know what you are taking about. Will make that change in a follow-up PR.

Comment (mbasmanova) [reply velox/exec/GroupingSet.cpp:335]: Submitted PR #5640

**Comment by aditi-pandit** [comment]
@mbasmanova : Just FYI... I wondered if window aggregates also could be used this way. But Presto doesn't support them yet. Nor does Trino : https://github.com/trinodb/trino/issues/7885

**Pull request closed without merging**